src/main/scala/ir/transforms/validate/Ackermann.scala

```diff
@@ -0,0 +1,404 @@
+      Some(
+        SideEffectStatement(
+          e,
+          s"Load_${endianHint(endian)}_${typeHint(addr.getType)}_${typeHint(lhs.getType)}",
+          rets,
+          args
+        )
+      )
+    case MemoryStore(memory, addr, value, endian, size, _) =>
+      val args = List(traceOut, Field("addr") -> addr)
+      val rets = List(traceOut)
+      Some(
+        SideEffectStatement(
+          e,
+          s"Store_${endianHint(endian)}_${typeHint(addr.getType)}_${typeHint(value.getType)}",
+          rets,
+          args
+        )
+      )
+    case MemoryAssign(lhs, rhs, _) =>
+      Some(
+        SideEffectStatement(
+          e,
+          s"MemoryAssign_${typeHint(lhs.getType)}_${typeHint(rhs.getType)}",
+          List(traceOut, Field("out") -> lhs),
+          List(traceOut, Field("arg") -> rhs)
+        )
+      )
+    case IndirectCall(arg, _) =>
+      // kind of want these gone :(
+      // doesn't capture interprocedural effects but we don't resolve indirect calls so it doesn't matter
+      Some(
+        SideEffectStatement(
+          e,
+          s"IndirectCall_${typeHint(arg.getType)}",
+          List(traceOut, Field("target") -> arg),
+          List(traceOut)
+        )
+      )
+    case _ => None
+  }
+
+}
+
+object Ackermann {
+
+  case class AckInv(name: String, lhs: List[CompatArg], rhs: List[CompatArg]) {
+    def toPredicate(renameSource: Expr => Expr, renameTarget: Expr => Expr) = {
+      val args = boolAnd(rhs.map { case CompatArg(s, t) =>
+        BinaryExpr(EQ, renameSource(s), renameTarget(t))
+      })
+      val implicant = boolAnd(lhs.map { case CompatArg(s, t) =>
+        BinaryExpr(EQ, renameSource(s), renameTarget(t))
+      })
+      BinaryExpr(BoolIMPLIES, args, implicant)
+    }
+  }
+
+  enum InstFailureReason {
+    case NameMismatch(msg: String)
+    case ParamMismatch(msg: String)
+  }
+
+  /**
+   * Check compatibility of two side effects and emit the lists (lhs, rhs) such that 
+   *
+   *  `(\forall (si, ti)  \in lhs si == ti) ==> (\forall (so, to) \in rhs . so == to)`
+   *
+   */
+  def instantiateAxiomInstance(
+    renaming: TransformDataRelationFun
+  )(source: SideEffectStatement, target: SideEffectStatement): Either[InstFailureReason, AckInv] = {
+    // source has higher level, has params, target does not have params
+
+    import InstFailureReason.*
+
+    def applyRename(a: EffCallFormalParam): EffCallFormalParam = a match {
+      case a: Field => a
+      case a: (Memory | Variable) =>
+        renaming(source.parent.parent.name, None)(a).toList match {
+          case (n: EffCallFormalParam) :: Nil => n
+          case (n: EffCallFormalParam) :: tl => n
+          case n :: Nil =>
+            tvLogger.warn(
+              s"Transform description fun rewrite formal parameter $a to $n, which I can't fit back into the formal parameter type Variable | Memory | Field, ignoring"
+            )
+            a
+          case Nil => a
```

Comment:
optionally, I think that you can use `List(x, y)` as a pattern match - if you want to avoid dealing with Nil